pkg/apis/core/validation/validation_test.go

```diff
@@ -10552,29 +10552,417 @@ func TestValidateServiceCreate(t *testing.T) {
 			},
 			numErrs: 1,
 		},
+		/* ip families validation */
+		{
+			name: "invalid, service with invalid ipFamilies",
+			tweakSvc: func(s *core.Service) {
+				invalidServiceIPFamily := core.IPFamily("not-a-valid-ip-family")
+				s.Spec.IPFamilies = []core.IPFamily{invalidServiceIPFamily}
+			},
+			numErrs: 1,
+		},
+		{
+			name: "invalid, service with invalid ipFamilies (2nd)",
+			tweakSvc: func(s *core.Service) {
+				invalidServiceIPFamily := core.IPFamily("not-a-valid-ip-family")
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, invalidServiceIPFamily}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, wrong IPFamilyPolicy(singleStack) is set for two families",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &singleStack
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "valid, IPFamilyPolicy(preferDualStack) is set for two families (note: alloc sets families)",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &preferDualStack
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol}
+			},
+			numErrs:              0,
+			dualStackFeatureGate: true,
+		},
+
+		{
+			name: "invalid, service with 2+ ipFamilies",
+			tweakSvc: func(s *core.Service) {
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol, core.IPv6Protocol, core.IPv4Protocol}
+			},
+			numErrs:              2,
+			dualStackFeatureGate: true,
+		},
 		{
-			name: "valid, nil service IPFamily",
+			name: "invalid, service with same ip families",
 			tweakSvc: func(s *core.Service) {
-				s.Spec.IPFamily = nil
+				s.Spec.IPFamilyPolicy = &requireDualStack
+				s.Spec.IPFamilies = []core.IPFamily{core.IPv6Protocol, core.IPv6Protocol}
+			},
+			numErrs:              1,
+			dualStackFeatureGate: true,
+		},
+		{
+			name: "invalid, dual stack service, dual stack feature gate is off ",
```

Comment:
I think this case is moot after the gate comments are resolved, and in fact the field in this struct can go away, too